crates/next-core/src/page_loader.rs

```diff
@@ -0,0 +1,153 @@
+            stringify_str(&format!("/{}", &*this.pathname.await?))
+        )?;
+
+        let base_code = next_js_file("entry/page-loader.ts");
+        if let FileContent::Content(base_file) = &*base_code.await? {
+            result += base_file.content()
+        } else {
+            bail!("required file `entry/page-loader.ts` not found");
+        }
+
+        let file = File::from(result.build());
+
+        Ok(VirtualAssetVc::new(this.entry_asset.path().join("page-loader.ts"), file.into()).into())
+    }
+
+    #[turbo_tasks::function]
+    async fn get_page_chunks(self) -> Result<ChunksVc> {
+        let this = &*self.await?;
+
+        let loader_entry_asset = self.get_loader_entry_asset();
+
+        let asset = EcmascriptModuleAssetVc::new(
+            loader_entry_asset.into(),
+            this.client_context,
+            turbo_tasks::Value::new(EcmascriptModuleAssetType::Typescript),
+            EcmascriptInputTransformsVc::cell(vec![EcmascriptInputTransform::TypeScript]),
+            this.client_context.environment(),
+        );
+
+        let chunk_group =
+            ChunkGroupVc::from_chunk(asset.as_evaluated_chunk(this.client_chunking_context, None));
+
+        Ok(chunk_group.chunks())
+    }
+}
+
+#[turbo_tasks::value_impl]
+impl Asset for PageLoaderAsset {
+    #[turbo_tasks::function]
+    async fn path(&self) -> Result<FileSystemPathVc> {
+        Ok(self
+            .server_root
+            .join("_next/static/chunks/pages")
+            .join(&get_asset_path_from_route(&*self.pathname.await?, ".js")))
+    }
+
+    #[turbo_tasks::function]
+    async fn content(self_vc: PageLoaderAssetVc) -> Result<AssetContentVc> {
+        let this = &*self_vc.await?;
+
+        let chunks = self_vc.get_page_chunks();
+
+        let mut data = Vec::new();
+        for chunk in chunks.await?.iter() {
+            let path = chunk.path().await?;
+            data.push(Value::String(format!("/{}", path.path)));
+        }
+
+        let content = format!(
+            "__turbopack_load_page_chunks__({}, {})\n",
+            stringify_str(&*this.pathname.await?),
+            Value::Array(data)
+        );
+        let file = File::from(content).with_content_type(APPLICATION_JAVASCRIPT);
+
+        Ok(AssetContentVc::from(file))
+    }
+
+    #[turbo_tasks::function]
+    async fn references(self_vc: PageLoaderAssetVc) -> Result<AssetReferencesVc> {
+        let chunks = self_vc.get_page_chunks();
+
+        let mut references = Vec::new();
+        for chunk in chunks.await?.iter() {
+            references.push(ChunkReferenceVc::new(*chunk).into());
+        }
+
+        Ok(AssetReferencesVc::cell(references))
+    }
+}
+
+fn get_asset_path_from_route(route: &str, ext: &str) -> String {
+    let path = if route == "" {
+        format!("index")
+    } else if route == "index" || route.starts_with("index/") {
+        format!("index/{}", route)
+    } else {
+        format!("{}", route)
+    };
+
+    return path + ext;
+}
```

Comment:
Your other PR has this in  `crates/next-core/src/util.rs`

https://github.com/vercel/turbo/pull/2949/files#diff-8c29575d9d107b60b83bd9b4449cf8cf46f4587ebd30a859db99c7bc3fb5d4d1R85